// @heroku-cli/plugin-apps/lib/commands/domains/add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const color_1 = require("@heroku-cli/color");
const command_1 = require("@heroku-cli/command");
const cli_ux_1 = require("cli-ux");
const shellescape = require("shell-escape");
const wait_for_domain_1 = require("../../lib/wait-for-domain");
class DomainsAdd extends command_1.Command {
    async run() {
        const { args, flags } = this.parse(DomainsAdd);
        const { hostname } = args;
        cli_ux_1.default.action.start(`Adding ${color_1.color.green(args.hostname)} to ${color_1.color.app(flags.app)}`);
        const { body: domain } = await this.heroku.post(`/apps/${flags.app}/domains`, {
            body: { hostname }
        });
        cli_ux_1.default.action.stop();
        if (flags.json) {
            cli_ux_1.default.styledJSON(domain);
        }
        else {
            cli_ux_1.default.log(`Configure your app's DNS provider to point to the DNS Target ${color_1.color.green(domain.cname || '')}.
For help, see https://devcenter.heroku.com/articles/custom-domains`);
            if (domain.status !== 'none') {
                if (flags.wait) {
                    await wait_for_domain_1.default(flags.app, this.heroku, domain);
                }
                else {
                    cli_ux_1.default.log('');
                    cli_ux_1.default.log(`The domain ${color_1.color.green(hostname)} has been enqueued for addition`);
                    let command = `heroku domains:wait ${shellescape([hostname])}`;
                    cli_ux_1.default.log(`Run ${color_1.color.cmd(command)} to wait for completion`);
                }
            }
        }
    }
}
DomainsAdd.description = 'add a domain to an app';
DomainsAdd.examples = ['heroku domains:add www.example.com'];
DomainsAdd.flags = {
    help: command_1.flags.help({ char: 'h' }),
    app: command_1.flags.app({ required: true }),
    json: command_1.flags.boolean({ description: 'output in json format', char: 'j' }),
    wait: command_1.flags.boolean()
};
DomainsAdd.args = [{ name: 'hostname' }];
exports.default = DomainsAdd;
